File: client/mq.go
package client

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"strings"
	"time"

	"github.com/google/uuid"
)

type MQData struct {
	Cmd       string `json:"cmd"`
	Topic     string `json:"topic"`
	Payload   string `json:"payload"`
	RequestId string `json:"requestId"`
	ReplayId  string `json:"replayId"`
	Regtopic  string `json:"regtopic"`
	Error     string `json:"error"`
	FromId    string `json:"fromId"`
}
type MQResponse struct {
	Payload string `json:"payload"`
	Error   string `json:"error"`
}

func jsonToStructResponse(data string) (*MQResponse, error) {
	var res MQResponse
	err := json.Unmarshal([]byte(data), &res)
	if err != nil {
		return nil, err
	}
	return &res, nil
}
func structToJSONResponse(res MQResponse) (string, error) {
	bytes, err := json.Marshal(res)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}
func jsonToStruct(data string) (*MQData, error) {
	var mq MQData
	err := json.Unmarshal([]byte(data), &mq)
	if err != nil {
		return nil, err
	}
	return &mq, nil
}
func structToJSON(mq MQData) (string, error) {
	bytes, err := json.Marshal(mq)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

type MQ struct {
	ID        string
	conn      net.Conn
	subs      map[string][]func(msg MQData)
	chs       map[string]chan string     // cria um canal de string
	chrequest map[string]chan MQResponse // cria um canal de string
	services  map[string]func(msg MQData, replay func(err string, payload string))
}

func Dial(url string) (*MQ, error) {

	info, err := ParseMQURL(url)
	if err != nil {
		return nil, err
	}
	conn, err := net.Dial("tcp", info.Host+":"+info.Port)
	if err != nil {
		return nil, err
	}
	mq := MQ{
		conn:      conn,
		chs:       make(map[string]chan string),
		chrequest: make(map[string]chan MQResponse),
		services:  map[string]func(msg MQData, replay func(err string, payload string)){},
		subs:      make(map[string][]func(msg MQData)),
	}
	mq.Send(MQData{
		Cmd:     "AUTH",
		Topic:   "",
		Payload: info.User + ":" + info.Pass,
	})
	go mq.on()
	return &mq, nil
}

func (mq *MQ) Send(data MQData) error {
	str, err := structToJSON(data)
	if err != nil {
		return err
	}
	_, err = mq.conn.Write([]byte(str + "\n"))
	return err
}

func (mq *MQ) Service(topic string, fn func(msg MQData, replay func(err string, payload string))) {
	mq.services[topic] = fn
	mq.Send(MQData{
		Cmd:     "SER",
		Topic:   topic,
		Payload: "",
	})
}

func (mq *MQ) Sub(topic string, cb func(msg MQData)) {
	mq.subs[topic] = append(mq.subs[topic], cb)
	mq.Send(MQData{
		Cmd:     "SUB",
		Topic:   topic,
		Payload: "",
	})
}

func (mq *MQ) Pub(topic, Payload string) {
	mq.Send(MQData{
		Cmd:     "PUB",
		Topic:   topic,
		Payload: Payload,
	})
}

func (mq *MQ) Req(topic, Payload string, timeout time.Duration) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "REQ",
		Topic:     topic,
		RequestId: reqId,
		Payload:   Payload,
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", topic)
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(timeout):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal %s", timeout, topic)
	}
}

func (mq *MQ) Ping() (string, error) {
	reqId := uuid.New().String()
	mq.chs[reqId] = make(chan string) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "PING",
		Topic:     "",
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chs[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		return res, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}

func (mq *MQ) Get(key string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "GET",
		Topic:     key,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) Set(key, value string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "SET",
		Topic:     key,
		RequestId: reqId,
		Payload:   value,
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}

func (mq *MQ) Del(key string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "DEL",
		Topic:     key,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) FilterKey(filter string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BFK",
		Topic:     "",
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) FilterValue(filter string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BFV",
		Topic:     "",
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) BFilterKey(bucket, filter string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BFK",
		Topic:     bucket,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) BFilterValue(bucket, filter string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BFV",
		Topic:     bucket,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) BGet(bucket, key string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "GET",
		Topic:     bucket + ":" + key,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) BSet(bucket, key, value string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "SET",
		Topic:     bucket + ":" + key,
		RequestId: reqId,
		Payload:   value,
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}

func (mq *MQ) BDel(bucket, key string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "DEL",
		Topic:     bucket + ":" + key,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}

func (mq *MQ) DeleteBucket(name string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BDEL",
		Topic:     name,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}

func (mq *MQ) CreateBucket(name string) (string, error) {
	reqId := uuid.New().String()
	mq.chrequest[reqId] = make(chan MQResponse) // cria um canal de string
	mq.Send(MQData{
		Cmd:       "BADD",
		Topic:     name,
		RequestId: reqId,
		Payload:   "",
	})
	ch, existe := mq.chrequest[reqId]
	if !existe {
		return "", fmt.Errorf("canal %s não existe", "")
	}

	select {
	case res := <-ch:
		close(ch)
		if res.Error != "" {
			return "", errors.New("Error :" + res.Error)
		}
		return res.Payload, nil
	case <-time.After(1 * time.Second):
		close(ch)
		return "", fmt.Errorf("timeout de %v expirado no canal", 1)
	}
}
func (mq *MQ) on() {
	reader := bufio.NewReader(mq.conn)
	for {
		// Lê a mensagem do cliente até encontrar uma nova linha
		str, err := reader.ReadString('\n')
		if err != nil {
			fmt.Printf("Erro ao ler: %s\n", err.Error())
			return
		}

		data, err := jsonToStruct(str)
		if err != nil {
			fmt.Printf("Erro ao ler: %s\n", err.Error())
			return
		}
		switch data.Cmd {
		case "CNN":
			mq.ID = data.Payload
			fmt.Println("connected")
			//fmt.Println(data)
		case "OK":
			//fmt.Println(data)
		case "ER_AUH":
			mq.Stop()
			panic(data.Payload)
			//fmt.Println(data)
		case "RES", "SET", "GET", "DEL", "BDEL", "BADD", "BFK", "BFV":
			ch, existe := mq.chrequest[data.RequestId]
			if !existe {
				return
			}

			ch <- MQResponse{
				Payload: data.Payload,
				Error:   data.Error,
			}

		case "PONG":
			ch, existe := mq.chs[data.RequestId]
			if !existe {
				return
			}
			ch <- data.Payload
		case "REQ":
			if mq.services[data.Topic] != nil {
				mq.services[data.Topic](*data, func(err string, payload string) {
					mq.Send(MQData{
						Cmd:       "RES",
						RequestId: data.RequestId,
						ReplayId:  data.ReplayId,
						Topic:     data.Topic,
						Error:     err,
						Payload:   payload,
					})
				})
			}
		case "PUB":
			topic := data.Topic
			if strings.Contains(data.Regtopic, ".*.") {
				topic = data.Regtopic
			}
			for _, sub := range mq.subs[topic] {
				go sub(*data)
			}
		}

	}
}

func (mq *MQ) Stop() error {
	return mq.conn.Close()
}

File: bun/mq.js
const net = require('net');
const crypto = require('node:crypto');

class MQData {
    constructor({ cmd, topic, payload, requestId, replayId, regtopic, error, fromId }) {
        this.cmd = cmd;
        this.topic = topic;
        this.payload = payload;
        this.requestId = requestId;
        this.replayId = replayId;
        this.regtopic = regtopic;
        this.error = error;
        this.fromId = fromId;
    }
}

class MQResponse {
    constructor({ payload, error }) {
        this.payload = payload;
        this.error = error;
    }
}

function jsonToStructResponse(data) {
    try {
        return new MQResponse(JSON.parse(data));
    } catch (err) {
        throw err;
    }
}

function structToJSONResponse(res) {
    try {
        return JSON.stringify(res);
    } catch (err) {
        throw err;
    }
}

function jsonToStruct(data) {
    try {
        return new MQData(JSON.parse(data));
    } catch (err) {
        throw err;
    }
}

function structToJSON(mq) {
    try {
        return JSON.stringify(mq);
    } catch (err) {
        throw err;
    }
}

class MQ {
    constructor() {
        this.ID = '';
        this.conn = null;
        this.subs = new Map();
        this.chs = new Map();
        this.chrequest = new Map();
        this.services = new Map();
    }

    static async connect(url) {
        const info = MQ.parseMQURL(url);
        const mq = new MQ();
        
        return new Promise((resolve, reject) => {
            const conn = net.createConnection({
                host: info.host,
                port: info.port
            }, () => {
                mq.conn = conn;
                mq.send(new MQData({
                    cmd: "AUTH",
                    topic: "",
                    payload: info.user + ":" + info.pass
                }));
                mq.on();
                resolve(mq);
            });
            
            conn.on('error', (err) => {
                reject(err);
            });
        });
    }

    static parseMQURL(url) {
        // Basic URL parsing - you might want to enhance this
        const parts = url.split('://')[1].split('@');
        const [user, pass] = parts[0].split(':');
        const [host, port] = parts[1].split(':');
        return { user, pass, host, port };
    }

    send(data) {
        return new Promise((resolve, reject) => {
            try {
                const str = structToJSON(data);
                this.conn.write(str + '\n', (err) => {
                    if (err) reject(err);
                    else resolve();
                });
            } catch (err) {
                reject(err);
            }
        });
    }

    service(topic, fn) {
        this.services.set(topic, fn);
        this.send(new MQData({
            cmd: "SER",
            topic: topic,
            payload: ""
        })).catch(console.error);
    }

    sub(topic, cb) {
        if (!this.subs.has(topic)) {
            this.subs.set(topic, []);
        }
        this.subs.get(topic).push(cb);
        this.send(new MQData({
            cmd: "SUB",
            topic: topic,
            payload: ""
        })).catch(console.error);
    }

    pub(topic, payload) {
        this.send(new MQData({
            cmd: "PUB",
            topic: topic,
            payload: payload
        })).catch(console.error);
    }

    async req(topic, payload, timeout = 1000) {
        const reqId = crypto.randomUUID();
        this.chrequest.set(reqId, {
            response: null,
            resolve: null,
            reject: null
        });

        await this.send(new MQData({
            cmd: "REQ",
            topic: topic,
            requestId: reqId,
            payload: payload
        }));

        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.chrequest.delete(reqId);
                reject(new Error(`timeout of ${timeout}ms expired for topic ${topic}`));
            }, timeout);

            this.chrequest.get(reqId).resolve = (res) => {
                clearTimeout(timer);
                this.chrequest.delete(reqId);
                if (res.error) {
                    reject(new Error("Error: " + res.error));
                } else {
                    resolve(res.payload);
                }
            };

            this.chrequest.get(reqId).reject = (err) => {
                clearTimeout(timer);
                this.chrequest.delete(reqId);
                reject(err);
            };
        });
    }

    async ping() {
        const reqId = crypto.randomUUID();
        this.chs.set(reqId, {
            response: null,
            resolve: null,
            reject: null
        });

        await this.send(new MQData({
            cmd: "PING",
            topic: "",
            requestId: reqId,
            payload: ""
        }));

        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.chs.delete(reqId);
                reject(new Error(`timeout of 1000ms expired for ping`));
            }, 1000);

            this.chs.get(reqId).resolve = (res) => {
                clearTimeout(timer);
                this.chs.delete(reqId);
                resolve(res);
            };

            this.chs.get(reqId).reject = (err) => {
                clearTimeout(timer);
                this.chs.delete(reqId);
                reject(err);
            };
        });
    }

    async get(key) {
        return this.req("GET", key);
    }

    async set(key, value) {
        return this.req("SET", value, 1000, key);
    }

    async del(key) {
        return this.req("DEL", "", 1000, key);
    }

    async filterKey(filter) {
        return this.req("BFK", "", 1000, filter);
    }

    async filterValue(filter) {
        return this.req("BFV", "", 1000, filter);
    }

    async bFilterKey(bucket, filter) {
        return this.req("BFK", "", 1000, bucket + ":" + filter);
    }

    async bFilterValue(bucket, filter) {
        return this.req("BFV", "", 1000, bucket + ":" + filter);
    }

    async bGet(bucket, key) {
        return this.req("GET", "", 1000, bucket + ":" + key);
    }

    async bSet(bucket, key, value) {
        return this.req("SET", value, 1000, bucket + ":" + key);
    }

    async bDel(bucket, key) {
        return this.req("DEL", "", 1000, bucket + ":" + key);
    }

    async deleteBucket(name) {
        return this.req("BDEL", "", 1000, name);
    }

    async createBucket(name) {
        return this.req("BADD", "", 1000, name);
    }

    on() {
        const reader = require('readline').createInterface({
            input: this.conn
        });

        reader.on('line', (str) => {
            try {
                const data = jsonToStruct(str);
                
                switch (data.cmd) {
                    case "CNN":
                        this.ID = data.payload;
                        console.log("connected");
                        break;
                    case "OK":
                        // Do nothing
                        break;
                    case "ER_AUH":
                        this.stop();
                        throw new Error(data.payload);
                    case "RES":
                    case "SET":
                    case "GET":
                    case "DEL":
                    case "BDEL":
                    case "BADD":
                    case "BFK":
                    case "BFV":
                        const ch = this.chrequest.get(data.requestId);
                        if (ch) {
                            ch.resolve(new MQResponse({
                                payload: data.payload,
                                error: data.error
                            }));
                        }
                        break;
                    case "PONG":
                        const pingCh = this.chs.get(data.requestId);
                        if (pingCh) {
                            pingCh.resolve(data.payload);
                        }
                        break;
                    case "REQ":
                        const service = this.services.get(data.topic);
                        if (service) {
                            service(data, (err, payload) => {
                                this.send(new MQData({
                                    cmd: "RES",
                                    requestId: data.requestId,
                                    replayId: data.replayId,
                                    topic: data.topic,
                                    error: err,
                                    payload: payload
                                })).catch(console.error);
                            });
                        }
                        break;
                    case "PUB":
                        let topic = data.topic;
                        if (data.regtopic && data.regtopic.includes('.*.')) {
                            topic = data.regtopic;
                        }
                        const subscribers = this.subs.get(topic) || [];
                        subscribers.forEach(sub => sub(data));
                        break;
                }
            } catch (err) {
                console.error(`Error processing message: ${err}`);
            }
        });

        reader.on('close', () => {
            console.log('Connection closed');
        });

        this.conn.on('error', (err) => {
            console.error(`Connection error: ${err}`);
        });

        this.conn.on('end', () => {
            console.log('Connection ended');
            reader.close();
        });
    }

    stop() {
        if (this.conn) {
            this.conn.end();
        }
    }
}

module.exports = {
    MQ,
    MQData,
    MQResponse
};